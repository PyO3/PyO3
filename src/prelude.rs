//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
