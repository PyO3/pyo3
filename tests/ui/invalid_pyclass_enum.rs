use pyo3::prelude::*;

#[pyclass(subclass)]
enum NotBaseClass {
    X,
    Y,
}

#[pyclass(extends = PyList)]
enum NotDrivedClass {
    X,
    Y,
}

#[pyclass]
enum NoEmptyEnum {}

#[pyclass]
enum NoUnitVariants {
    StructVariant { field: i32 },
    UnitVariant,
}

#[pyclass]
enum SimpleNoSignature {
    #[pyo3(constructor = (a, b))]
    A,
    B,
}

#[pyclass(eq, eq_int)]
enum SimpleEqOptRequiresPartialEq {
    A,
    B,
}

#[pyclass(eq)]
enum ComplexEqOptRequiresPartialEq {
    A(i32),
    B { msg: String },
}

#[pyclass(eq_int)]
enum SimpleEqIntWithoutEq {
    A,
    B,
}

#[pyclass(eq_int)]
enum NoEqInt {
    A(i32),
    B { msg: String },
}

#[pyclass(frozen, eq, eq_int, hash)]
#[derive(PartialEq)]
enum SimpleHashOptRequiresHash {
    A,
    B,
}

#[pyclass(frozen, eq, hash)]
#[derive(PartialEq)]
enum ComplexHashOptRequiresHash {
    A(i32),
    B { msg: String },
}

#[pyclass(hash)]
#[derive(Hash)]
enum SimpleHashOptRequiresFrozenAndEq {
    A,
    B,
}

#[pyclass(hash)]
#[derive(Hash)]
enum ComplexHashOptRequiresEq {
    A(i32),
    B { msg: String },
}

#[pyclass(ord)]
enum InvalidOrderedComplexEnum {
    VariantA (i32),
    VariantB { msg: String }
}

#[pyclass(eq,ord)]
#[derive(PartialEq)]
enum InvalidOrderedComplexEnum2 {
    VariantA (i32),
    VariantB { msg: String }
}

#[pyclass(eq)]
#[derive(PartialEq)]
enum AllEnumVariantsDisabled {
    #[cfg(any())]
    DisabledA,
    #[cfg(not(all()))]
    DisabledB,
}

fn main() {}
