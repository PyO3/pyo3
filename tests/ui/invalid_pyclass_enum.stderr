error: enums can't be inherited by other classes
 --> tests/ui/invalid_pyclass_enum.rs:3:11
  |
3 | #[pyclass(subclass)]
  |           ^^^^^^^^

error: enums can't extend from other classes
 --> tests/ui/invalid_pyclass_enum.rs:9:11
  |
9 | #[pyclass(extends = PyList)]
  |           ^^^^^^^

error: #[pyclass] can't be used on enums without any variants
  --> tests/ui/invalid_pyclass_enum.rs:16:18
   |
16 | enum NoEmptyEnum {}
   |                  ^^

error: Unit variant `UnitVariant` is not yet supported in a complex enum
       = help: change to an empty tuple variant instead: `UnitVariant()`
       = note: the enum is complex because of non-unit variant `StructVariant`
  --> tests/ui/invalid_pyclass_enum.rs:21:5
   |
21 |     UnitVariant,
   |     ^^^^^^^^^^^

error: `constructor` can't be used on a simple enum variant
  --> tests/ui/invalid_pyclass_enum.rs:26:12
   |
26 |     #[pyo3(constructor = (a, b))]
   |            ^^^^^^^^^^^

error: The `eq_int` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_enum.rs:43:11
   |
43 | #[pyclass(eq_int)]
   |           ^^^^^^

error: `eq_int` can only be used on simple enums.
  --> tests/ui/invalid_pyclass_enum.rs:49:11
   |
49 | #[pyclass(eq_int)]
   |           ^^^^^^

error: The `hash` option requires the `frozen` option.
  --> tests/ui/invalid_pyclass_enum.rs:69:11
   |
69 | #[pyclass(hash)]
   |           ^^^^

error: The `hash` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_enum.rs:69:11
   |
69 | #[pyclass(hash)]
   |           ^^^^

error: The `hash` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_enum.rs:76:11
   |
76 | #[pyclass(hash)]
   |           ^^^^

error: The `ord` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_enum.rs:83:11
   |
83 | #[pyclass(ord)]
   |           ^^^

error: #[pyclass] can't be used on enums without any variants - all variants of enum `AllEnumVariantsDisabled` have been configured out by cfg attributes
  --> tests/ui/invalid_pyclass_enum.rs:98:6
   |
98 | enum AllEnumVariantsDisabled {
   |      ^^^^^^^^^^^^^^^^^^^^^^^

error[E0369]: binary operation `==` cannot be applied to type `&SimpleEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:31:11
   |
31 | #[pyclass(eq, eq_int)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `SimpleEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:32:1
   |
32 | enum SimpleEqOptRequiresPartialEq {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `SimpleEqOptRequiresPartialEq` with `#[derive(PartialEq)]`
   |
32 + #[derive(PartialEq)]
33 | enum SimpleEqOptRequiresPartialEq {
   |

error[E0369]: binary operation `!=` cannot be applied to type `&SimpleEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:31:11
   |
31 | #[pyclass(eq, eq_int)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `SimpleEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:32:1
   |
32 | enum SimpleEqOptRequiresPartialEq {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `SimpleEqOptRequiresPartialEq` with `#[derive(PartialEq)]`
   |
32 + #[derive(PartialEq)]
33 | enum SimpleEqOptRequiresPartialEq {
   |

error[E0369]: binary operation `==` cannot be applied to type `&ComplexEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:37:11
   |
37 | #[pyclass(eq)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `ComplexEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:38:1
   |
38 | enum ComplexEqOptRequiresPartialEq {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `ComplexEqOptRequiresPartialEq` with `#[derive(PartialEq)]`
   |
38 + #[derive(PartialEq)]
39 | enum ComplexEqOptRequiresPartialEq {
   |

error[E0369]: binary operation `!=` cannot be applied to type `&ComplexEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:37:11
   |
37 | #[pyclass(eq)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `ComplexEqOptRequiresPartialEq`
  --> tests/ui/invalid_pyclass_enum.rs:38:1
   |
38 | enum ComplexEqOptRequiresPartialEq {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `ComplexEqOptRequiresPartialEq` with `#[derive(PartialEq)]`
   |
38 + #[derive(PartialEq)]
39 | enum ComplexEqOptRequiresPartialEq {
   |

error[E0277]: the trait bound `SimpleHashOptRequiresHash: Hash` is not satisfied
  --> tests/ui/invalid_pyclass_enum.rs:55:31
   |
55 | #[pyclass(frozen, eq, eq_int, hash)]
   |                               ^^^^ the trait `Hash` is not implemented for `SimpleHashOptRequiresHash`
   |
help: consider annotating `SimpleHashOptRequiresHash` with `#[derive(Hash)]`
   |
57 + #[derive(Hash)]
58 | enum SimpleHashOptRequiresHash {
   |

error[E0277]: the trait bound `ComplexHashOptRequiresHash: Hash` is not satisfied
  --> tests/ui/invalid_pyclass_enum.rs:62:23
   |
62 | #[pyclass(frozen, eq, hash)]
   |                       ^^^^ the trait `Hash` is not implemented for `ComplexHashOptRequiresHash`
   |
help: consider annotating `ComplexHashOptRequiresHash` with `#[derive(Hash)]`
   |
64 + #[derive(Hash)]
65 | enum ComplexHashOptRequiresHash {
   |

error[E0369]: binary operation `>` cannot be applied to type `&InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:89:14
   |
89 | #[pyclass(eq,ord)]
   |              ^^^
   |
note: an implementation of `PartialOrd` might be missing for `InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:91:1
   |
91 | enum InvalidOrderedComplexEnum2 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialOrd`
help: consider annotating `InvalidOrderedComplexEnum2` with `#[derive(PartialEq, PartialOrd)]`
   |
91 + #[derive(PartialEq, PartialOrd)]
92 | enum InvalidOrderedComplexEnum2 {
   |

error[E0369]: binary operation `<` cannot be applied to type `&InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:89:14
   |
89 | #[pyclass(eq,ord)]
   |              ^^^
   |
note: an implementation of `PartialOrd` might be missing for `InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:91:1
   |
91 | enum InvalidOrderedComplexEnum2 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialOrd`
help: consider annotating `InvalidOrderedComplexEnum2` with `#[derive(PartialEq, PartialOrd)]`
   |
91 + #[derive(PartialEq, PartialOrd)]
92 | enum InvalidOrderedComplexEnum2 {
   |

error[E0369]: binary operation `<=` cannot be applied to type `&InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:89:14
   |
89 | #[pyclass(eq,ord)]
   |              ^^^
   |
note: an implementation of `PartialOrd` might be missing for `InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:91:1
   |
91 | enum InvalidOrderedComplexEnum2 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialOrd`
help: consider annotating `InvalidOrderedComplexEnum2` with `#[derive(PartialEq, PartialOrd)]`
   |
91 + #[derive(PartialEq, PartialOrd)]
92 | enum InvalidOrderedComplexEnum2 {
   |

error[E0369]: binary operation `>=` cannot be applied to type `&InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:89:14
   |
89 | #[pyclass(eq,ord)]
   |              ^^^
   |
note: an implementation of `PartialOrd` might be missing for `InvalidOrderedComplexEnum2`
  --> tests/ui/invalid_pyclass_enum.rs:91:1
   |
91 | enum InvalidOrderedComplexEnum2 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialOrd`
help: consider annotating `InvalidOrderedComplexEnum2` with `#[derive(PartialEq, PartialOrd)]`
   |
91 + #[derive(PartialEq, PartialOrd)]
92 | enum InvalidOrderedComplexEnum2 {
   |
